File: src/share/make-n-gram.ts
import { range } from '../helper/iterable-helper'

export { makeWordNGram, makeCharNGram }

const makeNGram = (size: number, source: Array<string>): Array<Array<string>> => {
    const end = source.length - size + 1
    return range(end)
        .map(i => source.slice(i, i + size))
}

const makeWordNGram = (size: number, source: string): Array<Array<string>> => {
    return makeNGram(size, source.split(' '))
}

const makeCharNGram = (size: number, source: string): Array<string> => {
    return makeNGram(size, source.split('')).map(chars => chars.join(''))
}

File: src/share/make_n_gram.py
from typing import List


def make_n_gram(size: int, source: List[str]) -> List[List[str]]:
    end = len(source) - size + 1
    return list(map(lambda i: list(source[i : (i + size)]), range(end)))


def make_word_n_gram(size: int, source: str) -> List[List[str]]:
    return make_n_gram(size, source.split(" "))


def make_char_n_gram(size: int, source: str) -> List[str]:
    return list(map("".join, make_n_gram(size, list(source))))
